src/main/java/org/opensearch/security/privileges/ResourceAccessEvaluator.java

```diff
@@ -0,0 +1,212 @@
+     */
+    public PrivilegesEvaluatorResponse evaluate(
+        final ActionRequest request,
+        final String action,
+        boolean isResourceSharingFeatureEnabled,
+        final PrivilegesEvaluationContext context,
+        final PrivilegesEvaluatorResponse presponse
+    ) {
+
+        // If feature is disabled we skip evaluation through this evaluator
+        if (!isResourceSharingFeatureEnabled) {
+            return presponse;
+        }
+
+        // Skip evaluation if request is not a DocRequest type
+        if (!(request instanceof DocRequest req)) {
+            return presponse;
+        }
+
+        log.debug("Evaluating resource access");
+
+        // Resource Creation requests must be checked by regular index access evaluator
+        if (req.id() == null) {
+            log.debug("Request id is null, request is of type {}", req.getClass().getName());
+            return presponse;
+        }
+
+        // if requested index is not a resource sharing index, move on to the next evaluator
+        if (!resourceIndices.contains(req.index())) {
+            log.debug("Request index {} is not a protected resource index", req.index());
+            return presponse;
+        }
+
+        final UserSubjectImpl userSubject = (UserSubjectImpl) this.threadContext.getPersistent(
+            ConfigConstants.OPENDISTRO_SECURITY_AUTHENTICATED_USER
+        );
+        final User user = (userSubject == null) ? null : userSubject.getUser();
+
+        if (user == null) {
+            presponse.allowed = false;
+            log.debug("User is not authenticated, returning unauthorized");
+            return presponse.markComplete();
+        }
+
+        // If the user is a super-admin, the request would have already been granted. So no need to check whether user is admin.
+
+        CountDownLatch latch = new CountDownLatch(1);
+
+        Set<String> userRoles = new HashSet<>(user.getSecurityRoles());
+        Set<String> userBackendRoles = new HashSet<>(user.getRoles());
+
+        AtomicBoolean shouldMarkAsComplete = new AtomicBoolean(false);
+
+        // Fetch the ResourceSharing document and evaluate access
+        this.resourceSharingIndexHandler.fetchSharingInfo(req.index(), req.id(), ActionListener.wrap(document -> {
+            if (document == null) {
+                // TODO check whether we should mark response as not allowed. At present, it just returns incomplete response and hence is
+                // delegated to next evaluator
+                log.debug("No resource sharing record found for resource {} and index {}, skipping evaluation.", req.id(), req.index());
+                latch.countDown();
+                return;
+            }
+
+            // If the document is public, action is allowed
+            // If user is the owner, action is allowed
+            if (document.isSharedWithEveryone() || document.isCreatedBy(user.getName())) {
```

Comment:
Does `isSharedWithEveryone` means that everyone has full access or that its visible to everyone?